Use v2 api from distribution

dmcgowan opened the pull request:
Update registry package to use the v2 registry api from distribution. Update interfaces to directly take in digests.

dmcgowan committed:
Vendor distribution v2 api

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

jfrazelle commented:
i think this should wait until the registry is refactored no? this is a huge patch

dmcgowan commented:
The only non-vendor files touched is the registry/*.go and graph/*.go files, and the only non-refactor change is using the digest Parser instead of the string split.

stevvooe commented on graph/push.go:
```diff
@@ -378,13 +377,13 @@ func (s *TagStore) pushV2Repository(r *registry.Session, localRepo Repository, o
 
 			var exists bool
 			if len(checksum) > 0 {
-				sumParts := strings.SplitN(checksum, ":", 2)
```
It's good to see these random string splits going away.

stevvooe commented:
LGTM! Needs a rebase though.

dmcgowan committed:
Use vendored v2 registry api

Update registry package to use the v2 registry api from distribution. Update interfaces to directly take in digests.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
Rebased

> i think this should wait until the registry is refactored no? this is a huge patch

I do recall you requesting we break the distribution refactor into smaller PRs. This would be an example of such a PR, less than 50 lines of changes to graph/ and registry/ package, the remaining changes are related to moving and vendoring.

stevvooe commented:
@jfrazelle This is just using the API descriptors and errors from the distribution project. The refactor is a different set of work, focused on using a provided client.

jfrazelle commented:
oh sorry missed the whole vendor change thing going on here, carry on

dmcgowan commented:
Carry on as move from design review to code review? :grinning:

jfrazelle commented:
oh idk about that

On Tue, Mar 31, 2015 at 4:31 PM, Derek McGowan <notifications@github.com>
wrote:

> Carry on as move from design review to code review? [image: :grinning:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11478#issuecomment-88283226>.
>

jfrazelle commented:
need another maintainer eyes to approve first :angel:

On Tue, Mar 31, 2015 at 4:32 PM, Jessica Frazelle <jess@docker.com> wrote:

> oh idk about that
>
> On Tue, Mar 31, 2015 at 4:31 PM, Derek McGowan <notifications@github.com>
> wrote:
>
>> Carry on as move from design review to code review? [image: :grinning:]
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/docker/docker/pull/11478#issuecomment-88283226>.
>>
>
>

dmp42 commented:
ping @tiborvass @icecrime @LK4D4 @vbatts and the cat :)

xiaods commented:
can't wait for this feature. +1

icecrime commented:
Cool! LGTM

vbatts commented:
LGTM

tiborvass commented:
LGTM tested with private registry v1 and v2
